src/control/control.rs

```diff
@@ -0,0 +1,173 @@
+            Keystate::Press(tim) => {
+                if now > tim + KEYHOLDDELAY {
+                    self.d = Keystate::Hold;
+                }
+            }
+            _ => {}
+        }
+        match self.l {
+            Keystate::Press(tim) => {
+                if now > tim + KEYHOLDDELAY {
+                    self.l = Keystate::Hold;
+                }
+            }
+            _ => {}
+        }
+        match self.r {
+            Keystate::Press(tim) => {
+                if now > tim + KEYHOLDDELAY {
+                    self.r = Keystate::Hold;
+                }
+            }
+            _ => {}
+        }
+        
+        let dir = match (self.u, self.d, self.l, self.r) {
+            (Keystate::Hold | Keystate::Tap, Keystate::None, Keystate::None, Keystate::None) => Some((0, -1)),
+            (Keystate::None, Keystate::Hold | Keystate::Tap, Keystate::None, Keystate::None) => Some((0, 1)),
+            (Keystate::None, Keystate::None, Keystate::Hold | Keystate::Tap, Keystate::None) => Some((-1, 0)),
+            (Keystate::None, Keystate::None, Keystate::None, Keystate::Hold | Keystate::Tap) => Some((1, 0)),
+            
+            (Keystate::Press(_) | Keystate::Hold | Keystate::Tap, Keystate::None, Keystate::Press(_) | Keystate::Hold | Keystate::Tap, Keystate::None) => Some((-1, -1)),
+            (Keystate::Press(_) | Keystate::Hold | Keystate::Tap, Keystate::None, Keystate::None, Keystate::Press(_) | Keystate::Hold | Keystate::Tap) => Some((1, -1)),
+            (Keystate::None, Keystate::Press(_) | Keystate::Hold | Keystate::Tap, Keystate::Press(_) | Keystate::Hold | Keystate::Tap, Keystate::None) => Some((-1, 1)),
+            (Keystate::None, Keystate::Press(_) | Keystate::Hold | Keystate::Tap, Keystate::None, Keystate::Press(_) | Keystate::Hold | Keystate::Tap) => Some((1, 1)),
+
+            _ => None,
+        };
+
+        if self.u == Keystate::Tap {
+            self.u = Keystate::None;
+        }
+        if self.d == Keystate::Tap {
+            self.d = Keystate::None;
+        }
+        if self.l == Keystate::Tap {
+            self.l = Keystate::None;
+        }
+        if self.r == Keystate::Tap {
+            self.r = Keystate::None;
+        }
+        
+        match dir {
+            Some(dir) => {
+                callstack.push(gamedata.players[pid].lock().unwrap().class.mov(pid, dir, now));
+                match self.u {
+                    Keystate::Press(_) => { self.u = Keystate::Hold; }
+                    _ => {}
+                }
+                match self.d {
+                    Keystate::Press(_) => { self.d = Keystate::Hold; }
+                    _ => {}
+                }
+                match self.l {
+                    Keystate::Press(_) => { self.l = Keystate::Hold; }
+                    _ => {}
+                }
+                match self.r {
+                    Keystate::Press(_) => { self.r = Keystate::Hold; }
+                    _ => {}
```

Comment:
All of these would be better served by `if let` than with `match`

Reply:
I actually think for this and the previous one, you should make a closure that uses `if let` and takes in a mutable reference to a value (i.e. `&mut self.u`) and assigns to the dereferencing of that, which would allow you to eliminate this code repetition.